rust/kernel/delay.rs

```diff
@@ -0,0 +1,35 @@
+// SPDX-License-Identifier: GPL-2.0
+
+//! Delay functions for operations like sleeping.
+//!
+//! C header: [`include/linux/delay.h`](../../../../include/linux/delay.h)
+
+use crate::bindings;
+use core::{cmp::min, time::Duration};
+
+/// Sleeps safely even with waitqueue interruptions.
+///
+/// This function forwards the call to the C side `msleep` function. As a result,
+/// `duration` will be rounded up to the nearest millisecond if granularity less
+/// than a millisecond is provided. Any [`Duration`] that exceeds
+/// [`c_uint::MAX`][core::ffi::c_uint::MAX] in milliseconds is saturated.
+pub fn coarse_sleep(duration: Duration) {
+    let duration_as_nanos = duration.as_nanos();
```

Comment:
What kind of code is generated for `u128`? I think it's preferable to avoid these types, especially because we know the result needs to fit in a `u32` anyway.

Reply:
`as_nanos` does `self.secs as u128 * NANOS_PER_SEC as u128 + self.nanos as u128` internally. You can use `.as_secs()` and `.subsec_nanos()` to get the internal fields of the `Duration`.

Reply:
I've removed usage of `u128`s.